<a href="https://colab.research.google.com/github/narenderkulariya/st20269341-CMP7005-PRAC1/blob/main/st20269341_CMP7005_PRAC1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Introduction**

**Task 1: Data Handling**

• Download the dataset India air quality dataset, in CSV format, from Moodle.

• Import the dataset into your development environment (e.g., Jupyter Notebook,Colab).

• Perform initial inspection and merge the datasets into one to initiate a comprehensive analysis.

# Importing the required **libraries**

In [3]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os
import glob
import numpy as np
import seaborn as sns
from sklearn import preprocessing
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline
import plotly.express as px
import plotly.graph_objects as go
import seaborn as sns

# Mounting the **drive**

In [4]:
# Loading the Drive helper and mounting the drive
from google.colab import drive

# using prompt for authorization.
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
%cd '/content/drive/MyDrive/pds asset data set'
# locating the data path

/content/drive/MyDrive/pds asset data set


In [6]:
%ls # it helps to show all the content of folder

Ahmedabad_data.csv       Chennai_data.csv     Kochi_data.csv
Aizawl_data.csv          Coimbatore_data.csv  Kolkata_data.csv
all_cities_combined.csv  Delhi_data.csv       Lucknow_data.csv
Amaravati_data.csv       Ernakulam_data.csv   Mumbai_data.csv
Amritsar_data.csv        Gurugram_data.csv    Patna_data.csv
Bengaluru_data.csv       Guwahati_data.csv    Shillong_data.csv
Bhopal_data.csv          Hyderabad_data.csv   Talcher_data.csv
Brajrajnagar_data.csv    Jaipur_data.csv      Thiruvananthapuram_data.csv
Chandigarh_data.csv      Jorapokhar_data.csv  Visakhapatnam_data.csv


In [7]:
# Import the necessary libraries
import pandas as pd  # pandas is used for working with data tables
import glob         # glob is used to find files by name patterns

# STEP 1: The pattern "*_data.csv" means "find all files that end with '_data.csv'"
# This will find files like: Ahmedabad_data.csv, Delhi_data.csv, Mumbai_data.csv, etc.
city_files = glob.glob("*_data.csv")

# STEP 2: Create an empty list to store all our city data
# We'll put each city's data in this list before combining them
all_cities_data = []

# STEP 3: Read each city file one by one
for file_name in city_files:
    # Read the current city's CSV file into a DataFrame
    # A DataFrame is like a spreadsheet table in Python
    city_df = pd.read_csv(file_name)

    # Add this city's data to our list
    all_cities_data.append(city_df)

    # Optional: Print which file we just read
    print(f"Loaded: {file_name}")

# STEP 4: Combine all city data into one big table
# pd.concat() joins all the DataFrames in our list together
# ignore_index=True makes sure the row numbers are continuous (0, 1, 2, 3...)
combined_data = pd.concat(all_cities_data, ignore_index=True)

# STEP 5: Save the combined data to a new CSV file
# index=False means don't save the row numbers as a separate column
combined_data.to_csv("all_cities_combined.csv", index=False)

# STEP 6: Show us what we accomplished
# len(city_files) = count of how many city files we combined
# len(combined_data) = total number of rows in the final combined file
print(f"SUCCESS: Combined {len(city_files)} city files into one file with {len(combined_data)} total rows")
print("The combined file is saved as: all_cities_combined.csv")

Loaded: Ahmedabad_data.csv
Loaded: Amritsar_data.csv
Loaded: Mumbai_data.csv
Loaded: Kochi_data.csv
Loaded: Kolkata_data.csv
Loaded: Bengaluru_data.csv
Loaded: Patna_data.csv
Loaded: Guwahati_data.csv
Loaded: Hyderabad_data.csv
Loaded: Ernakulam_data.csv
Loaded: Chennai_data.csv
Loaded: Chandigarh_data.csv
Loaded: Gurugram_data.csv
Loaded: Brajrajnagar_data.csv
Loaded: Coimbatore_data.csv
Loaded: Thiruvananthapuram_data.csv
Loaded: Visakhapatnam_data.csv
Loaded: Lucknow_data.csv
Loaded: Jaipur_data.csv
Loaded: Shillong_data.csv
Loaded: Amaravati_data.csv
Loaded: Aizawl_data.csv
Loaded: Delhi_data.csv
Loaded: Bhopal_data.csv
Loaded: Jorapokhar_data.csv
Loaded: Talcher_data.csv
SUCCESS: Combined 26 city files into one file with 29531 total rows
The combined file is saved as: all_cities_combined.csv


In [8]:
df= pd.read_csv('all_cities_combined.csv')
df

City        Date  PM2.5   PM10     NO    NO2    NOx    NH3     CO  \
0      Ahmedabad  01/01/2015    NaN    NaN   0.92  18.22  17.15    NaN   0.92   
1      Ahmedabad  02/01/2015    NaN    NaN   0.97  15.69  16.46    NaN   0.97   
2      Ahmedabad  03/01/2015    NaN    NaN  17.40  19.30  29.70    NaN  17.40   
3      Ahmedabad  04/01/2015    NaN    NaN   1.70  18.48  17.97    NaN   1.70   
4      Ahmedabad  05/01/2015    NaN    NaN  22.10  21.42  37.76    NaN  22.10   
...          ...         ...    ...    ...    ...    ...    ...    ...    ...   
29526    Talcher  27/06/2020  20.75  25.94   8.17  21.95  15.89   6.73   1.18   
29527    Talcher  28/06/2020  24.88  31.07    NaN    NaN   0.00    NaN   1.55   
29528    Talcher  29/06/2020  20.37  25.46  22.82  19.64  18.96  10.23   1.59   
29529    Talcher  30/06/2020  18.62  23.26  21.24  26.53  47.60   4.45   1.30   
29530    Talcher  01/07/2020  24.85  31.06  27.73  29.36  57.09   1.37   1.52   

         SO2      O3  Benzene  Toluene  Xylene    AQI    AQI_Bucket  
0      27.64  133.36     0.00     0.02    0.00    NaN           NaN  
1      24.55   34.06     3.68     5.50    3.77    NaN           NaN  
2      29.07   30.70     6.80    16.40    2.25    NaN           NaN  
3      18.59   36.08     4.43    10.14    1.00    NaN           NaN  
4      39.33   39.31     7.01    18.89    2.78    NaN           NaN  
...      ...     ...      ...      ...     ...    ...           ...  
29526  18.62   48.87     0.00      NaN     NaN   90.0  Satisfactory  
29527  21.33   48.55     0.00      NaN     NaN   98.0  Satisfactory  
29528  20.79   46.84     0.00      NaN     NaN   99.0  Satisfactory  
29529  14.34   28.77     0.00      NaN     NaN   86.0  Satisfactory  
29530  14.42   14.90     0.00      NaN     NaN  101.0      Moderate  

[29531 rows x 16 columns]

**Perform Fundamental Data Understanding :**

In [9]:
df.head()

City        Date  PM2.5  PM10     NO    NO2    NOx  NH3     CO    SO2  \
0  Ahmedabad  01/01/2015    NaN   NaN   0.92  18.22  17.15  NaN   0.92  27.64   
1  Ahmedabad  02/01/2015    NaN   NaN   0.97  15.69  16.46  NaN   0.97  24.55   
2  Ahmedabad  03/01/2015    NaN   NaN  17.40  19.30  29.70  NaN  17.40  29.07   
3  Ahmedabad  04/01/2015    NaN   NaN   1.70  18.48  17.97  NaN   1.70  18.59   
4  Ahmedabad  05/01/2015    NaN   NaN  22.10  21.42  37.76  NaN  22.10  39.33   

       O3  Benzene  Toluene  Xylene  AQI AQI_Bucket  
0  133.36     0.00     0.02    0.00  NaN        NaN  
1   34.06     3.68     5.50    3.77  NaN        NaN  
2   30.70     6.80    16.40    2.25  NaN        NaN  
3   36.08     4.43    10.14    1.00  NaN        NaN  
4   39.31     7.01    18.89    2.78  NaN        NaN

In [10]:
df.tail()

City        Date  PM2.5   PM10     NO    NO2    NOx    NH3    CO  \
29526  Talcher  27/06/2020  20.75  25.94   8.17  21.95  15.89   6.73  1.18   
29527  Talcher  28/06/2020  24.88  31.07    NaN    NaN   0.00    NaN  1.55   
29528  Talcher  29/06/2020  20.37  25.46  22.82  19.64  18.96  10.23  1.59   
29529  Talcher  30/06/2020  18.62  23.26  21.24  26.53  47.60   4.45  1.30   
29530  Talcher  01/07/2020  24.85  31.06  27.73  29.36  57.09   1.37  1.52   

         SO2     O3  Benzene  Toluene  Xylene    AQI    AQI_Bucket  
29526  18.62  48.87      0.0      NaN     NaN   90.0  Satisfactory  
29527  21.33  48.55      0.0      NaN     NaN   98.0  Satisfactory  
29528  20.79  46.84      0.0      NaN     NaN   99.0  Satisfactory  
29529  14.34  28.77      0.0      NaN     NaN   86.0  Satisfactory  
29530  14.42  14.90      0.0      NaN     NaN  101.0      Moderate

In [11]:
df.shape
print(f'No of Rows: {df.shape[0]}, No of Columns: {df.shape[1]}')

No of Rows: 29531, No of Columns: 16


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


In [13]:
df.dtypes

City           object
Date           object
PM2.5         float64
PM10          float64
NO            float64
NO2           float64
NOx           float64
NH3           float64
CO            float64
SO2           float64
O3            float64
Benzene       float64
Toluene       float64
Xylene        float64
AQI           float64
AQI_Bucket     object
dtype: object

In [14]:
df.columns

Index(['City', 'Date', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2',
       'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket'],
      dtype='object')

**Number of cities in the dataset**

In [15]:
cities = df['City'].value_counts()
print(f'Total number of cities in the dataset : {len(cities)}')
cities

Total number of cities in the dataset : 26


City
Ahmedabad             2009
Mumbai                2009
Bengaluru             2009
Delhi                 2009
Lucknow               2009
Chennai               2009
Hyderabad             2006
Patna                 1858
Gurugram              1679
Visakhapatnam         1462
Amritsar              1221
Jorapokhar            1169
Jaipur                1114
Thiruvananthapuram    1112
Amaravati              951
Brajrajnagar           938
Talcher                925
Kolkata                814
Guwahati               502
Coimbatore             386
Shillong               310
Chandigarh             304
Bhopal                 289
Kochi                  162
Ernakulam              162
Aizawl                 113
Name: count, dtype: int64

**Converting the Date column into Date and Time format**

In [16]:
# convert Date to datetime format as in the original dataframe the type of date column is object
df['Date'] = pd.to_datetime(df['Date'], format="%d/%m/%Y")
df['month']=pd.DatetimeIndex(df['Date']).month #making a new column by extracting month part of the Date column in cities dataframe
df['Year']=pd.DatetimeIndex(df['Date']).year ##making a new column by extracting year part of the Date column in cities dataframe



In [17]:
df.head()

City       Date  PM2.5  PM10     NO    NO2    NOx  NH3     CO    SO2  \
0  Ahmedabad 2015-01-01    NaN   NaN   0.92  18.22  17.15  NaN   0.92  27.64   
1  Ahmedabad 2015-01-02    NaN   NaN   0.97  15.69  16.46  NaN   0.97  24.55   
2  Ahmedabad 2015-01-03    NaN   NaN  17.40  19.30  29.70  NaN  17.40  29.07   
3  Ahmedabad 2015-01-04    NaN   NaN   1.70  18.48  17.97  NaN   1.70  18.59   
4  Ahmedabad 2015-01-05    NaN   NaN  22.10  21.42  37.76  NaN  22.10  39.33   

       O3  Benzene  Toluene  Xylene  AQI AQI_Bucket  month  Year  
0  133.36     0.00     0.02    0.00  NaN        NaN      1  2015  
1   34.06     3.68     5.50    3.77  NaN        NaN      1  2015  
2   30.70     6.80    16.40    2.25  NaN        NaN      1  2015  
3   36.08     4.43    10.14    1.00  NaN        NaN      1  2015  
4   39.31     7.01    18.89    2.78  NaN        NaN      1  2015

In [18]:
df.tail()

City       Date  PM2.5   PM10     NO    NO2    NOx    NH3    CO  \
29526  Talcher 2020-06-27  20.75  25.94   8.17  21.95  15.89   6.73  1.18   
29527  Talcher 2020-06-28  24.88  31.07    NaN    NaN   0.00    NaN  1.55   
29528  Talcher 2020-06-29  20.37  25.46  22.82  19.64  18.96  10.23  1.59   
29529  Talcher 2020-06-30  18.62  23.26  21.24  26.53  47.60   4.45  1.30   
29530  Talcher 2020-07-01  24.85  31.06  27.73  29.36  57.09   1.37  1.52   

         SO2     O3  Benzene  Toluene  Xylene    AQI    AQI_Bucket  month  \
29526  18.62  48.87      0.0      NaN     NaN   90.0  Satisfactory      6   
29527  21.33  48.55      0.0      NaN     NaN   98.0  Satisfactory      6   
29528  20.79  46.84      0.0      NaN     NaN   99.0  Satisfactory      6   
29529  14.34  28.77      0.0      NaN     NaN   86.0  Satisfactory      6   
29530  14.42  14.90      0.0      NaN     NaN  101.0      Moderate      7   

       Year  
29526  2020  
29527  2020  
29528  2020  
29529  2020  
29530  2020

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 18 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   City        29531 non-null  object        
 1   Date        29531 non-null  datetime64[ns]
 2   PM2.5       24933 non-null  float64       
 3   PM10        18391 non-null  float64       
 4   NO          25949 non-null  float64       
 5   NO2         25946 non-null  float64       
 6   NOx         25346 non-null  float64       
 7   NH3         19203 non-null  float64       
 8   CO          27472 non-null  float64       
 9   SO2         25677 non-null  float64       
 10  O3          25509 non-null  float64       
 11  Benzene     23908 non-null  float64       
 12  Toluene     21490 non-null  float64       
 13  Xylene      11422 non-null  float64       
 14  AQI         24850 non-null  float64       
 15  AQI_Bucket  24850 non-null  object        
 16  month       29531 non-

**Looking at the missing values in the dataset**

In [20]:
# Missing values
def missing_values_table(df):
        # Total missing values
        print("\nMissing Values Count:")
        mis_val = df.isnull().sum()

        # Percentage of missing values
        mis_val_percent = 100 * df.isnull().sum() / len(df)

        # Make a table with the results
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        print(mis_val_table)

        # Rename the columns
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})

        # Sort the table by percentage of missing descending
        mis_val_table_ren_columns = mis_val_table_ren_columns.sort_values(
        '% of Total Values', ascending=False)

        # Return the dataframe with missing information
        return mis_val_table_ren_columns

missing_values= missing_values_table(df)
missing_values.style.background_gradient(cmap='Purples')


Missing Values Count:
                0          1
City            0   0.000000
Date            0   0.000000
PM2.5        4598  15.570079
PM10        11140  37.723071
NO           3582  12.129626
NO2          3585  12.139785
NOx          4185  14.171549
NH3         10328  34.973418
CO           2059   6.972334
SO2          3854  13.050692
O3           4022  13.619586
Benzene      5623  19.041008
Toluene      8041  27.229014
Xylene      18109  61.322001
AQI          4681  15.851139
AQI_Bucket   4681  15.851139
month           0   0.000000
Year            0   0.000000


## Total number of cities in the dataset:

In [21]:
cities = df['City'].value_counts()
print(f'Total number of cities in the dataset : {len(cities)}')
cities

Total number of cities in the dataset : 26


City
Ahmedabad             2009
Mumbai                2009
Bengaluru             2009
Delhi                 2009
Lucknow               2009
Chennai               2009
Hyderabad             2006
Patna                 1858
Gurugram              1679
Visakhapatnam         1462
Amritsar              1221
Jorapokhar            1169
Jaipur                1114
Thiruvananthapuram    1112
Amaravati              951
Brajrajnagar           938
Talcher                925
Kolkata                814
Guwahati               502
Coimbatore             386
Shillong               310
Chandigarh             304
Bhopal                 289
Kochi                  162
Ernakulam              162
Aizawl                 113
Name: count, dtype: int64

In [22]:
df['City'].unique()

array(['Ahmedabad', 'Amritsar', 'Mumbai', 'Kochi', 'Kolkata', 'Bengaluru',
       'Patna', 'Guwahati', 'Hyderabad', 'Ernakulam', 'Chennai',
       'Chandigarh', 'Gurugram', 'Brajrajnagar', 'Coimbatore',
       'Thiruvananthapuram', 'Visakhapatnam', 'Lucknow', 'Jaipur',
       'Shillong', 'Amaravati', 'Aizawl', 'Delhi', 'Bhopal', 'Jorapokhar',
       'Talcher'], dtype=object)

In [23]:
df['Year'].unique()

array([2015, 2016, 2017, 2018, 2019, 2020], dtype=int32)

In [24]:
df['month'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=int32)

In [25]:
# check any duplicated entries
print(f"There are  {df.duplicated().sum()} duplicates in the dataset")

There are  0 duplicates in the dataset


# Data Understanding **Results**

**Dataset Size:** The dataset started with 29,531 rows (records) and 16 columns (fields). After adding two new columns called 'Month' and 'Year', the total number of columns increased to 18.

**Data Types:** The dates in the dataset have been properly formatted as dates. The other information in the dataset is mostly stored as either numbers (for things like measurements) or text (for things like city names and categories).

**Missing Information:** Some columns have incomplete data. The most problematic columns are:

**Xylene:** Missing about 61% of its data

**PM10:** Missing about 38% of its data

**NH3:** Missing about 35% of its data

**Toluene:** Missing about 27% of its data

**AQI and AQI_Bucket:** Missing about 16% of their data

**Number of Cities:** The dataset includes information from 26 different cities.

**Time Range:** The data covers a 6-year period from 2015 to 2020, with measurements from all 12 months of each year.

**Duplicate Data:** The dataset is clean with no duplicate records found — each row contains unique information.

# **Imputing missing values**

# **Mean/Median/Mode Imputation**

**Mean:** Replace missing values with the mean of the column.

**Median:** Replace missing values with the median of the column. This is useful if your data has outliers, as the median is less sensitive to them.

**Mode:** Replace missing values with the mode (most frequent value) of the column. This is often used for categorical variables but can be applied to numerical data as well.

In [29]:
# pollutants = ['PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2',
#        'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket']
# dff= df.groupby(['Year','month','City'])[pollutants].mean().reset_index()
from sklearn.impute import SimpleImputer

pollutant_cols = [
    'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3',
    'CO', 'SO2', 'O3', 'Benzene', 'Toluene',
    'Xylene', 'AQI'
]

mean_imputer = SimpleImputer(strategy='mean')

df[pollutant_cols] = mean_imputer.fit_transform(df[pollutant_cols])


In [32]:
def classify_aqi(aqi):
    if pd.isna(aqi):
        return None
    elif aqi <= 50:
        return 'Good'
    elif aqi <= 100:
        return 'Satisfactory'
    elif aqi <= 200:
        return 'Moderate'
    elif aqi <= 300:
        return 'Poor'
    elif aqi <= 400:
        return 'Very Poor'
    else:
        return 'Severe'

df['AQI_Bucket'] = df['AQI'].apply(classify_aqi)

In [33]:
df.isnull().sum()

City          0
Date          0
PM2.5         0
PM10          0
NO            0
NO2           0
NOx           0
NH3           0
CO            0
SO2           0
O3            0
Benzene       0
Toluene       0
Xylene        0
AQI           0
AQI_Bucket    0
month         0
Year          0
dtype: int64

**Subsetting the columns**

Even though a lot of columns have been provided in the dataset, we shall select a few prominent ones. Let's create a new dataframe called pollutants containg the major pollutants responsible for air pollution.

# Handling Missing Values:

In [ ]:
df[['PM2.5','PM10','SO2','NO2','CO','O3','TEMP','PRES','DEWP','RAIN','wd','WSPM']] = df[['PM2.5','PM10','SO2','NO2','CO','O3','TEMP','PRES','DEWP','RAIN','wd','WSPM']].ffill()

Inference:
The forward-fill (ffill) method is used because it is well-suited for time-series environmental data like air quality and weather measurements, where readings are recorded continuously over time. In such datasets, missing values often occur due to short-term sensor malfunctions or data transmission issues. Using ffill replaces each missing value with the most recent valid observation, preserving the natural temporal continuity of the data. This approach prevents sudden breaks in the data trend and avoids introducing artificial fluctuations that other imputation methods (like mean or interpolation) might cause. Overall, ffill helps maintain a realistic and stable representation of environmental conditions over time.

Merging the year, month and day column and converting it into data time format

In [ ]:
# Merging year, month, day, and hour into a single datetime column
df['Date'] = pd.to_datetime(df[['year', 'month', 'day', 'hour']], format ='mixed')

# Drop the original year, month, and day columns
#df.drop(['year', 'month', 'day'], axis=1, inplace=True)
df.head()

In [ ]:
# Reordering the columns to place 'datetime' at first index position
columns = df.columns.tolist()
columns.insert(0, columns.pop(columns.index('Date')))
df = df[columns]
df

In [ ]:
print("Total missing values per Station:")
print(df.groupby('station').apply(lambda x: x.isnull().sum().sum()))